### PR TITLE
support build different python wheel in setup.py

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- Add a class named `SetupSpec` to support different python wheel in setup.py. 
- Build "ray-cpp" wheel when you set env "export RAY_INSTALL_CPP=1"
- If users want to use ray with C++ API, run "pip install ray[cpp]"
    * If you don't use "[cpp]", you will get an error when run "ray cpp --show-library-path"
- Add pip option "--find-links .whl/" in docker to ensure "ray-cpp" wheel can be found by "pip install ray[all]"

### CODE REVIEW

**Review by qicosmos — APPROVED** [review]

File: python/requirements.txt [hunk 1]
```diff
@@ -39,6 +39,7 @@ tensorboardX >= 1.9
 uvicorn
 dataclasses; python_version < '3.7'
 starlette
+ray-cpp
```
Comment (simon-mo) [inline python/requirements.txt:42]: don't add it here. this is for setting up the surrounding requirements

File: python/setup.py [hunk 2]
```diff
@@ -442,13 +465,15 @@ def has_ext_modules(self):
             return True
 
 
+# Ensure no remaining lib files.
+shutil.rmtree(os.path.join(ROOT_DIR, "build"))
```
Comment (simon-mo) [inline python/setup.py:469]: when do we generate this build dir?

File: python/requirements.txt (hunk 1 above)
Comment (SongGuyang) [reply python/requirements.txt:42]: ”ray-cpp“ package is a requirement of "ray" package in the way of "pip install ray[cpp]"?

File: python/setup.py (hunk 2 above)
Comment (SongGuyang) [reply python/setup.py:469]: "setup.py" will generate "build" dir but don't delete it. And if you rerun this script, you will include a lot of old files from "build" dir.

File: python/requirements.txt (hunk 1 above)
Comment (simon-mo) [reply python/requirements.txt:42]: this requirements.txt is not used to specify dependency. this file is used only to setup CI infra environment.

File: python/setup.py (hunk 2 above)
Comment (simon-mo) [reply python/setup.py:469]: ok got it

**Review by simon-mo — APPROVED** [review]

File: python/requirements.txt (hunk 1 above)
Comment (simon-mo) [reply python/requirements.txt:42]: so in order to CI to pass, you will need to remove this line

Comment (SongGuyang) [reply python/requirements.txt:42]: Got it!

File: python/setup.py [hunk 3]
```diff
@@ -21,12 +21,30 @@
 
 logger = logging.getLogger(__name__)
 
+
+class SetupSpec:
+    def __init__(self, name, description):
```
Comment (ijrsvt) [inline python/setup.py:26]: Can you add type hints here (for the ctor) and for the instance variables?

Comment (SongGuyang) [reply python/setup.py:26]: Make sense.

File: docker/ray-deps/Dockerfile [hunk 4]
```diff
@@ -3,11 +3,15 @@ FROM rayproject/base-deps:nightly"$GPU"
 # If this arg is not "autoscaler" then no autoscaler requirements will be included
 ARG AUTOSCALER="autoscaler"
 ARG WHEEL_PATH
+ARG FIND_LINKS_PATH
```
Comment (ijrsvt) [inline docker/ray-deps/Dockerfile:6]: Could you have these default to `ARG FIND_LINKS_PATH=".whl"`
(for all Dockerfiles)

File: python/setup.py [hunk 5]
```diff
@@ -489,11 +499,12 @@ def has_ext_modules(self):
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
     ],
-    packages=setuptools.find_packages() if setup_spec.name == "ray" else [],
+    packages=setuptools.find_packages()
+    if setup_spec.type == SetupType.RAY else [],
```
Comment (ijrsvt) [inline python/setup.py:503]: Could we make this a method fo SetupSpec? 

I.e.
```
class SetupSpec:
  def get_packages(self):
    if self.type == SetupType.RAY:
        return setuptools.find_packages()
    else:
        return []
```

File: python/setup.py [hunk 6]
```diff
@@ -60,6 +42,43 @@ def __init__(self, name, description):
...
+
+class SetupSpec:
+    def __init__(self, type, name, description):
+        self.type = type
+        self.name = name
+        self.version = find_version("ray", "__init__.py")
+        self.description = description
+        self.files_to_include = list()
+        self.install_requires = list()
+        self.extras = {}
```
Comment (ijrsvt) [inline python/setup.py:69]: Type hints

**Comment by simon-mo** [comment]
@SongGuyang can we make sure to test this end to end, maybe in https://github.com/ray-project/ray/blob/3380b68b54ce2dbfe1d6f2b7a1cf98815774e32f/.buildkite/pipeline.yml#L1-L21

**Comment by SongGuyang** [comment]
@simon-mo I found the label "ferris_wheel: Wheels and Jars" will upload all the wheels in directory "./.whl". And the "ray-cpp" wheels also locates in this directory. if I understand correctly, I don't need to add this test?

File: python/setup.py (hunk 5 above)
Comment (SongGuyang) [reply python/setup.py:503]: Good~

File: docker/ray-deps/Dockerfile (hunk 4 above)
Comment (SongGuyang) [reply docker/ray-deps/Dockerfile]: OK

**Comment by SongGuyang** [comment]
@simon-mo @ijrsvt Please help to see this PR agent. And do you know the best way to add end to end test? I want a test like this:
- pip  install --find-links .whl/ .whl/ray-2.0.0.dev0-cp38-cp38-linux_x86_64.whl[cpp]
- mkdir  /tmp/ray-demo && ray cpp --generate-bazel-project-template-to /tmp/ray-demo
- cd  /tmp/ray-demo && bazel run //:example
If I add this test to "docker/ray-deps/Dockerfile", does it make sense?

**Pull request merged**